pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+            A map between the edges and the number of keys required by the
+            edges
+        :param bool flexible: Dettermines if flexible can be use.
+            If False global settings will be attempted
         :raise PacmanRouteInfoAllocationException:
         """
         # check that this algorithm supports the constraints put onto the
         # partitions
-        self.__application_graph = application_graph
         self.__machine_graph = machine_graph
         self.__n_keys_map = n_keys_map
+        self.__n_bits_atoms_and_mac = 0
+        self.__n_bits_machine = 0
+        self.__n_bits_atoms = 0
+        self.__atom_bits_per_app_part = dict()
+        self.__flexible = flexible
+        self.__fixed_keys = list()
+        self.__fixed_partitions = dict()
+        self.__fixed_used = set()
 
         check_algorithm_can_support_constraints(
             constrained_vertices=machine_graph.outgoing_edge_partitions,
-            supported_constraints=[ContiguousKeyRangeContraint],
+            supported_constraints=[
+                ContiguousKeyRangeContraint, FixedKeyAndMaskConstraint],
             abstract_constraint_type=AbstractKeyAllocatorConstraint)
 
-        max_app_keys_bits, key_bits_per_app = self._calculate_zones()
+        self.__find_fixed()
+        self.__calculate_zones()
+        self.__check_zones()
+        return self.__allocate()
 
-        return self._simple_allocate(max_app_keys_bits, key_bits_per_app)
+    def __find_fixed(self):
+        by_app_and_partition_name = \
+            self.__machine_graph.multicast_partitions
+        for app_id in by_app_and_partition_name:
+            by_app = by_app_and_partition_name[app_id]
+            for partition_name, by_partition_name in by_app.items():
+                for mac_vertex in by_partition_name:
+                    partition = self.__machine_graph.\
+                        get_outgoing_edge_partition_starting_at_vertex(
+                            mac_vertex, partition_name)
+                    for constraint in partition.constraints:
+                        if isinstance(constraint, FixedKeyAndMaskConstraint):
+                            self.__add_fixed(partition, constraint)
 
-    def _calculate_zones(self):
+    def __add_fixed(self, partition, constraint):
+        self.__fixed_partitions[partition] = constraint.keys_and_masks
+        for key_and_mask in constraint.keys_and_masks:
+            # Go through the mask sets and save keys and n_keys
+            for key, n_keys in get_key_ranges(
+                    key_and_mask.key, key_and_mask.mask):
+                self.__fixed_keys.append((key, n_keys))
+
+    def __calculate_zones(self):
         """
-        :return: tuple containing max app keys and a map of app to max keys
-            for any given machine vertex
-        :rtype: tuple(int, dict(ApplicationVertex, int))
         :raises PacmanRouteInfoAllocationException:
         """
+        by_app_and_partition_name = \
```

Comment:
Is that simpler if it is just "machine_partitions"?

Reply:
I added a comment first tie it is used.